src/document/reducer.ts

```diff
@@ -299,34 +346,41 @@ export function baseReducer<T, A extends Action, L>(
     newDocument = produce(newDocument, draft => {
         // the reducer runs on a immutable version of
         // provided state
-
         try {
             const returnedDraft = customReducer(
                 draft.state,
                 _action as A,
                 dispatch,
             );
 
-            const clipboardValue = isUndoRedo(action) ? [...clipboard] : [];
+            // const clipboardValue = isUndoRedo(action) ? [...clipboard] : [];
 
             // if the reducer creates a new state object instead
             // of mutating the draft then returns the new state
             if (returnedDraft) {
                 // casts new state as draft to comply with typescript
                 return castDraft<Document<T, A, L>>({
                     ...newDocument,
-                    clipboard: [...clipboardValue],
+                    // clipboard: [...clipboardValue],
                     state: returnedDraft,
                 });
             } else {
-                draft.clipboard = castDraft([...clipboardValue]);
+                // draft.clipboard = castDraft([...clipboardValue]);
             }
         } catch (error) {
+            // TODO: if the reducer throws an error then we should keep the previous state (before replayOperations)
+            // and remove skip number from action/operation
             const lastOperationIndex =
                 newDocument.operations[_action.scope].length - 1;
             draft.operations[_action.scope][lastOperationIndex].error = (
                 error as Error
             ).message;
+
+            draft.operations[_action.scope][lastOperationIndex].skip = 0;
+
+            if (shouldProcessSkipOperation) {
+                draft.state = castDraft<State<T, L>>({ ...document.state });
+            }
         }
     });
 
```

Comment:
On an undo action, we can reuse the hash of the last non-skipped operation right? If we set it in the action then we avoid a `hashDocument` call in line 398:
```
        // if reuseHash is true, checks if the action has
        // an hash and uses it instead of generating it
        const scope = _action.scope || 'global';
        const hash =
            reuseHash && Object.prototype.hasOwnProperty.call(_action, 'hash')
                ? (_action as Operation).hash
                : hashDocument(draft, scope);
```